entry_types/scrolled/package/src/frontend/PlayerControls/WaveformPlayerControls/Waveform.js

```diff
@@ -0,0 +1,57 @@
+import React, {useRef} from 'react';
+import { blankSources } from "pageflow/frontend";
+import Measure from 'react-measure';
+import Wavesurfer from './Wavesurfer';
+
+import styles from './Waveform.module.css';
+
+const waveColor = '#828282ed';
+const waveColorInverted = 'rgba(0, 0, 0, 0.5)';
+
+const cursorColor = '#fff';
+const cursorColorInverted = '#888';
+
+export function Waveform(props) {
+  let isReady = false;
+  let mediaElementId = null;
+  let measureRect = useRef();
+
+  if (props.mediaElementId) {
+    mediaElementId = `#${props.mediaElementId}`;
+    let mediaElement = global.document.querySelector(mediaElementId);
+    if (mediaElement) {
+      if(mediaElement.src !== blankSources['audio'])
+        isReady = true;
+    }
+  }
```

Comment:
**issue**: Exception when adding or removing caption

When adding a caption, the component gets rerendered since it is now wrapped in a `Figure` component. During this rerender I observe a case where the media element is still in the document while this render function, but by the time the new `Waversurfer` component is mounted the media element has been removed. This causes a "Media element not found" exception in `_loadMediaElt` below.

The root of the problem, is using DOM APIs in a render function here. React render function should be pure. Since they potentially run a lot of time, doing more expensive stuff like looking up DOM nodes also risks causing performance issues.

I think, I have found a different solution. Which I will push to this branch along with the SSR fix. Maybe you can test again if it also covers all cases that you found during development

Reply:
I pushed two commits to fix SSR issue and issue mentioned here.